longbits/bit_builder.go

```diff
@@ -0,0 +1,178 @@
+//
+// Copyright 2019 Insolar Technologies GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package longbits
+
+import (
+	"math/bits"
+)
+
+func NewBitBuilder(expectedLen int) BitBuilder {
+	if expectedLen == 0 {
+		return BitBuilder{}
+	}
+	return BitBuilder{bytes: make([]byte, 0, expectedLen)}
+}
+
+type BitBuilder struct {
+	bytes       []byte
+	accumulator uint16
+}
+
+func (p BitBuilder) IsZero() bool {
+	return p.accumulator == 0 && p.bytes == nil
+}
+
+func (p BitBuilder) Len() int {
+	return (1+len(p.bytes))<<3 - bits.LeadingZeros8(uint8(p.accumulator>>8))
+}
+
+const accInit = 0x80
+
+func (p BitBuilder) ensure() {
+	if p.accumulator == 0 {
+		if len(p.bytes) != 0 {
+			panic("illegal state")
+		}
+		p.accumulator = accInit
+	}
+}
+
+func (p BitBuilder) Append(bit bool) BitBuilder {
+	p.ensure()
+
+	p.accumulator <<= 1
+	if bit {
+		p.accumulator |= 1
+	}
+	if p.accumulator >= accInit<<8 {
+		p.bytes = append(p.bytes, byte(p.accumulator))
+		p.accumulator = accInit
+	}
+
+	return p
+}
+
+func (p BitBuilder) AppendN(bitCount int, bit bool) BitBuilder {
+	p.ensure()
+
```

Comment:
todo - check for less then zero and panic

Reply:
fixed